### DISCUSSION TITLE
How to send or publish gRPC binary format data by using RAFT cluster?

### DISCUSSION THREAD

**Discussion by Jeevananthan-23** [dotnet/dotNext#152] [Q&A] [answered]
Using `Raft` consensus to create a gRPC-based primary server and send the data through replications.

**Comment by sakno** [comment, marked as answer]
What you mean by send or publish gRPC binary format? gRPC binary format is Protobuf. Protobuf can be used standalone as serialization format without gRPC. .NEXT Raft implementation has no restrictions over binary format of log entries stored in WAL and replicated across the cluster. Thus, you can use Protobuf, MessagePack, BSON, JSON, or whatever you like.

**Reply by Jeevananthan-23** [reply]
Is there any unit test for the same?

**Reply by sakno** [reply]
Especially for Protobuf? No. No need for that. There are a lot of tests for WAL and binary formatting of log entries. I can recommend to read [this](https://dotnet.github.io/dotNext/features/cluster/wal.html) article about WAL first. It fully describes programming model provided by the library to work with WAL. Also, take a look at [Raft playground](https://dotnet.github.io/dotNext/features/cluster/raft.html#example) to explore basic implementation of replication of binary log entries.

**Reply by sakno** [reply]
.NEXT Raft exposes low-level API to work with log entry contents. To be more precise, API exposes a way to work and transform raw byte buffers. The buffer can contain any kind of data - UTF-8 encoded JSON, custom binary message, i.e. everything that can be represented by a sequence of bytes.
